### PR TITLE
CORE-8956 Re-write url import in Go

### PR DESCRIPTION
**Pull request by sri2k1us**

### CODE REVIEW

File: urlimport.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	return u.Scheme, host, port, username, password, u.Path, 0
+}
+
+//DownloadFromURL downloads file from give url
+func DownloadFromURL() int {
+	response, err := http.Head(downloadURL)
+	if err != nil {
+		logMessage("Unable to complete HEAD request. " + err.Error())
```
Comment (ianmcorvidae) [inline urlimport.go]: This might be worth testing with some real-world cases. I know that e.g. apache and nginx and similar will always support `HEAD`, but it's not always implemented for more programmatically-generated sorts of things. At a guess, `curl` doesn't automatically do this, just goes right for a `GET`. So we might want to make sure adding a `HEAD` request doesn't break anyone's workflows.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,7 @@
 # urlimport
 A Go tool to download files from Http(s) and (s)ftp sites.
+- Support for Http(s) and (S)Ftp schemes
```
Comment (ianmcorvidae) [inline README.md]: Is this true? There's some acronym confusion about what exactly "SFTP" means, but I don't see anything in here or in the ftp library to do TLS dialing (maybe more commonly called "ftps", not sure? https://en.wikipedia.org/wiki/FTPS) nor do I see anything for supporting the SSH-based stuff often called SFTP (https://en.wikipedia.org/wiki/SSH_File_Transfer_Protocol).

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+FROM alpine
+COPY urlimport /bin
+ENTRYPOINT ["urlimport"]
```
Comment (ianmcorvidae) [inline Dockerfile]: (comment in a basically arbitrary place, so not really all that related to this line, though it does affect this file)

We'll probably want to set up a Jenkinsfile for this repo which tests and builds this. Most of our other go-based services use a Dockerfile which is set with a `FROM` line that's our golang-base image or at least one of the `golang:${version}-alpine` images, and do the building of the binary inside the container. They usually also install `github.com/jstemmer/go-junit-report` (it's in the golang-base image, or can be installed with `go get`) in order to produce junit-style output for Jenkins to parse. The Jenkinsfile for `saved-searches` is one of our go services which runs tests, so it might be a good place to look if you'd like; with an appropriate Dockerfile and a service.properties file you should be able to use that Jenkinsfile unchanged; it's written to be general and just pull the interesting parts out of that service.properties file.

File: urlimport.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+			os.Remove(logfileName)
+		} else {
+			os.Remove(filename) //there were some errors logged. Remove partial downloaded file.
+		}
+	}
+}
+
+//logMessage to a log file
+func logMessage(message string) {
+	log.WriteString(message)
```
Comment (ianmcorvidae) [inline urlimport.go:75]: Small thing, but should this write a newline too? I don't see newlines in what's passed to this function.

File: urlimport.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+//parseArgs parse command line args
+func parseArgs() int {
+	flag.StringVar(&filename, "filename", "", "file name to use when saving the imported file")
+	flag.StringVar(&downloadURL, "url", "", "Url to import the file from")
+	flag.Parse()
```
Comment (ianmcorvidae) [inline urlimport.go:83]: This is possibly unneeded, but I like the way the `flag` library is used in e.g. `de-job-killer`. In that construction, you specify the flag-related stuff when declaring the variable. I think it returns a pointer, which would change how some things are called, but probably not in too annoying a way.

**Review by ianmcorvidae — COMMENTED** [review]
A variety of small comments/thoughts scattered throughout; basically looks good to me though, just some potential edge cases and such given it's a relatively big change away from using curl.

File: README.md (hunk 2 above)
Comment (sri2k1us) [reply README.md]: Ok So both `sftp://demo:password@test.rebex.net:990/pub/example/readme.txt` and `sftp://demo:password@test.rebex.net:990/pub/example/readme.txt` are marked as error in UI itself in current DE url import . DE UI validates for `http(s) and ftp`. So I think this program will support what DE is currently supporting from the UI. I will update the `README.md` file.

File: urlimport.go (hunk 5 above)
Comment (sri2k1us) [reply urlimport.go:83]: 👍  Will take a look at the `de-job-killer`

File: urlimport.go (hunk 4 above)
Comment (sri2k1us) [reply urlimport.go:75]: 👍

File: Dockerfile (hunk 3 above)
Comment (sri2k1us) [reply Dockerfile]: Cool. I will take a look at `saved-searches`.

File: urlimport.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+import "net/http"
+import "os"
+import "io"
+import "io/ioutil"
+import "net/url"
+import "strings"
+import "time"
+import "net"
+
+import "github.com/jlaffaye/ftp"
```
Comment (slr71) [inline urlimport.go]: This is a minor thing, but most Go code that I've seen groups all of the imports in a single import statement:

```
import (
    "fmt"
    "flag"
    "net/http"
    "os"
    "io"
    "io/ioutil"
    "net/url"
    "strings"
    "time"
    "net"

    "github.com/jlaffaye/ftp"
)
```

File: urlimport.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+var filename string
+var downloadURL string
+var log *os.File
+var logfileName string
```
Comment (slr71) [inline urlimport.go]: This is another minor thing, but the variables defined here could be combined into a single `var` statement as well.

File: urlimport.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	log.WriteString("\n")
+}
+
+//parseArgs parse command line args
+func parseArgs() int {
+	flag.StringVar(&filename, "filename", "", "file name to use when saving the imported file")
+	flag.StringVar(&downloadURL, "url", "", "Url to import the file from")
+	flag.Parse()
+	fmt.Println(filename)
+	fmt.Println(downloadURL)
```
Comment (slr71) [inline urlimport.go]: Is this debugging code that should be removed?

**Review by slr71 — APPROVED** [review]
I had a couple of minor comments, but nothing important. It looks good to me. :shipit:

File: urlimport.go (hunk 6 above)
Comment (sri2k1us) [reply urlimport.go]: 👍

File: urlimport.go (hunk 7 above)
Comment (sri2k1us) [reply urlimport.go]: 👍

File: urlimport.go (hunk 8 above)
Comment (sri2k1us) [reply urlimport.go]: 👍
